mava/systems/gpo/anakin/rec_magpo.py

```diff
@@ -0,0 +1,804 @@
+    """Get the learner function."""
+
+    # Get apply functions for executing and training the network.
+    sable_action_select_fn, sable_apply_fn, actor_apply_fn = apply_fns
+    sable_update_fn, actor_update_fn = update_fn
+    num_envs = config.arch.num_envs
+    alpha = config.system.alpha
+    n_agents = env.num_agents
+    
+
+    def _update_step(learner_state: LearnerState, _: Any) -> Tuple[LearnerState, Tuple]:
+        """A single update of the network.
+
+        This function steps the environment and records the trajectory batch for
+        training. It then calculates advantages and targets based on the recorded
+        trajectory and updates the actor and critic networks based on the calculated
+        losses.
+
+        Args:
+        ----
+            learner_state (NamedTuple):
+                - params (FrozenDict): The current model parameters.
+                - opt_states (OptState): The current optimizer states.
+                - key (PRNGKey): The random number generator state.
+                - env_state (State): The environment state.
+                - last_timestep (TimeStep): The last timestep in the current trajectory.
+                - hstates (HiddenStates): The hidden state of the network.
+            _ (Any): The current metrics info.
+
+        """ 
+
+
+        def _env_step(
+            learner_state: LearnerState, _: Any
+        ) -> Tuple[LearnerState, Tuple[Transition, Metrics]]:
+            """Step the environment."""
+            params, opt_states, key, env_state, last_timestep, last_done, last_hstates = learner_state
+
+            # Select action
+            key, policy_key = jax.random.split(key)
+
+            # Apply the actor network to get the action, log_prob, value and updated hstates.
+            last_obs = last_timestep.observation
+            action, log_prob, value, sable_hstates = sable_action_select_fn(  # type: ignore
+                params.guider_params,
+                last_obs,
+                last_hstates.sable_hidden_state,
+                policy_key,
+            )
+
+            # Step environment
+            env_state, timestep = jax.vmap(env.step, in_axes=(0, 0))(env_state, action)
+
+            # Reset hidden state if done.
+            done = timestep.last()
+            last_done = done.repeat(env.num_agents).reshape(config.arch.num_envs, -1)
+            done = jnp.expand_dims(done, (1, 2, 3, 4))
+            sable_hstates = tree.map(lambda hs: jnp.where(done, jnp.zeros_like(hs), hs), sable_hstates)
+
+            prev_done = last_timestep.last().repeat(env.num_agents).reshape(num_envs, -1)
+            transition = Transition(
+                prev_done, action, value, timestep.reward, log_prob, last_timestep.observation, last_hstates
+            )
+            hstates = HiddenStates_all(sable_hstates, last_hstates.policy_hidden_state)
+            learner_state = LearnerState(params, opt_states, key, env_state, timestep, last_done, hstates)
```

Comment:
last_done/prev_done mean the same thing to me, I just made it more explicit with curr_done/prev_done
```suggestion
            done = timestep.last()
            done = jnp.expand_dims(done, (1, 2, 3, 4))
            sable_hstates = tree.map(lambda hs: jnp.where(done, jnp.zeros_like(hs), hs), sable_hstates)

            curr_done = done.repeat(env.num_agents).reshape(config.arch.num_envs, -1)
            prev_done = last_timestep.last().repeat(env.num_agents).reshape(num_envs, -1)
            transition = Transition(
                prev_done, action, value, timestep.reward, log_prob, last_timestep.observation, last_hstates
            )
            hstates = HiddenStates_all(sable_hstates, last_hstates.policy_hidden_state)
            learner_state = LearnerState(params, opt_states, key, env_state, timestep, curr_done, hstates)
```